src/System.Memory/src/System/Text/EncodingExtensions.cs

```diff
@@ -0,0 +1,649 @@
+        /// <param name="writer">The buffer to which the encoded bytes will be written.</param>
+        /// <exception cref="EncoderFallbackException">Thrown if <paramref name="chars"/> contains data that cannot be encoded and <paramref name="encoding"/> is configured
+        /// to throw an exception when such data is seen.</exception>
+        public static long GetBytes(this Encoding encoding, ReadOnlySpan<char> chars, IBufferWriter<byte> writer)
+        {
+            if (encoding is null)
+            {
+                throw new ArgumentNullException(nameof(encoding));
+            }
+
+            if (writer is null)
+            {
+                throw new ArgumentNullException(nameof(writer));
+            }
+
+            if (chars.Length <= MaxInputElementsPerIteration)
+            {
+                // The input span is small enough where we can one-shot this.
+
+                int byteCount = encoding.GetByteCount(chars);
+                Span<byte> scratchBuffer = writer.GetSpan(byteCount);
+
+                int actualBytesWritten = encoding.GetBytes(chars, scratchBuffer);
+                Debug.Assert(actualBytesWritten == byteCount, "Wrote unexpected number of bytes. Was buffer mutated unexpectedly?");
+
+                writer.Advance(actualBytesWritten);
+                return actualBytesWritten;
+            }
+            else
+            {
+                // Allocate a stateful Encoder instance and chunk this.
+
+                Convert(encoding.GetEncoder(), chars, writer, flush: true, out long totalBytesWritten, out bool completed);
+
+                Debug.Assert(completed, "Encoder should not have any intermediate state left over.");
+                return totalBytesWritten;
+            }
+        }
+
+        /// <summary>
+        /// Decodes the specified <see cref="ReadOnlySequence{Char}"/> to <see langword="byte"/>s using the specified <see cref="Encoding"/>
+        /// and writes the result to <paramref name="writer"/>.
+        /// </summary>
+        /// <param name="encoding">The <see cref="Encoding"/> which represents how the data in <paramref name="chars"/> should be encoded.</param>
+        /// <param name="chars">The <see cref="ReadOnlySequence{Char}"/> whose contents should be encoded.</param>
+        /// <param name="writer">The buffer to which the encoded bytes will be written.</param>
+        /// <returns>The number of bytes written to <paramref name="writer"/>.</returns>
+        /// <exception cref="EncoderFallbackException">Thrown if <paramref name="chars"/> contains data that cannot be encoded and <paramref name="encoding"/> is configured
+        /// to throw an exception when such data is seen.</exception>
+        public static long GetBytes(this Encoding encoding, in ReadOnlySequence<char> chars, IBufferWriter<byte> writer)
+        {
+            if (encoding is null)
+            {
+                throw new ArgumentNullException(nameof(encoding));
```

Comment:
Move the arg checking to another method? Did we stop doing that now? @stephentoub

Reply:
See https://github.com/dotnet/corefx/pull/41810/#discussion_r335753634.